org.metaborg.spoofax.eclipse/src/main/java/org/metaborg/spoofax/eclipse/editor/completion/SpoofaxCompletionProposal.java

```diff
@@ -71,37 +97,156 @@ public CompletionData(IDocument document, int offset, ICompletion completion) {
+
+            return result;
+
+        }
+
+        public int countPlaceholders(String input) {
+            int numberOfPlaceholders = 0;
+            for(int i = 0; i < input.length(); i++) {
+                if(input.charAt(i) == '[') {
+                    i++;
+                    if(input.charAt(i) == '[') {
+                        i++;
+
+                        while(i < input.length() && input.charAt(i) == '[')
+                            // nested brackets
+                            i++;
 
+                        while(i < input.length() && input.charAt(i) != ']') {
+                            String charAti = String.valueOf(input.charAt(i));
+
+                            if(!charAti.matches("[a-zA-Z_]")) { // not placeholder: abort
+                                break;
+                            }
+
+                            i++;
+                        }
+
+                        if(i >= input.length() || input.charAt(i) != ']')
+                            continue;
+                        i++;
+
+                        numberOfPlaceholders++;
+                        continue;
+                    }
+                }
+            }
+            return numberOfPlaceholders;
+        }
+
+        public int countPlaceholders(ICompletion completion) {
+            int numberOfPlaceholders = 0;
+            for(ICompletionItem i : completion.items()) {
+                if(i instanceof PlaceholderCompletionItem)
+                    numberOfPlaceholders++;
+            }
+            return numberOfPlaceholders;
+
+        }
+    }
+
+    private static final ILogger logger = LoggerUtils.logger(SpoofaxCompletionProposal.class);
     private final ITextViewer textViewer;
-    private final int offset;
-    private final ICompletion completion;
+    private int offset;
+    private ICompletion completion;
+    private final ICompletionService completionService;
+    private final ISyntaxService<?> syntaxService;
+    private final FileObject source;
+    private final ILanguageImpl language;
 
     private CompletionData data;
 
-
-    public SpoofaxCompletionProposal(ITextViewer textViewer, int offset, ICompletion completion) {
+    public SpoofaxCompletionProposal(ITextViewer textViewer, int offset, ICompletion completion, FileObject source,
+        ILanguageImpl language, ICompletionService completionService, ISyntaxService<?> syntaxService) {
         this.textViewer = textViewer;
         this.offset = offset;
         this.completion = completion;
+        this.completionService = completionService;
+        this.syntaxService = syntaxService;
+        this.source = source;
+        this.language = language;
     }
 
-
     @Override public void apply(IDocument document) {
-        this.data = new CompletionData(document, offset, completion);
+
+        int startOffset;
+        int endOffset;
+
+        // if completion is nested, replace the selected text
+        if(((Completion) completion).isNested()) {
```

Comment:
Is the cast to `Completion` needed there, or should `isNested()` be lifted into `ICompletion`?